### ISSUE TITLE
The ability to reload diagnostic

### ISSUE THREAD

**Issue by Shatur** [hrsh7th/vim-lamp#23]
Sometimes diagnostics are not properly updated. It would be nice to have the ability to trigger diagnostics update manually. Is it possible to implement?

**Comment by hrsh7th** [comment]
It's possible but I want to fix the problem...

It is most difficult critical bug of the `vim-lamp` now.

**Comment by hrsh7th** [comment]
@Shatur95 Currently, vim-lamp has an incremental diff feature that improves performance but I doubt the reason for the problem is it... maybe.

So I introduce the flag to toggle the feature.
Could you test to reproduce the problem with temporarily disable it?

**Comment by hrsh7th** [comment]
Could you test with `let g:lamp_experimental_passthrough_diagnostics = v:true
`?

When the flags turns on, you get a bit ugly behavior but it needs to investigate sorry...

**Comment by Shatur** [comment]
Good idea! I will try to provide minimal reproducible example and test new diagnostics feature.

**Comment by hrsh7th** [comment]
If you like a new feature than an old feature, please tell me it.

I don't know the best visualization for the diagnostics.

**Comment by Shatur** [comment]
What difference between `v:true` and `v:false` for `g:lamp_experimental_passthrough_diagnostics`?

**Comment by Shatur** [comment]
I tested with and without incremental feature. The problem still exits.

I found minimal reproducible example:
1. Download my [akd](https://github.com/Shatur95/akd)
2. Then you will need to generate `compile_commands.json`:

```vim
mkdir build
cd build
cmake -D CMAKE_EXPORT_COMPILE_COMMANDS ..
cp compile_commands.json ..
```

(you will have `compile_commands.json` in project code directory. I use a plugin for automation but I decided do describe manual steps)

3. Install and setup `clangd` in `vim-lamp`.
4. Open `keyboarddaemon.h` and create the following function `void test()`
5. Open `keyboarddaemon.cpp` and create the following definition `void KeyboardDaemon() {}`

**Expected behavior**

No errors detected.

**Actual behavior**

Diagnostics say that the function does not exists even after `:wa`.

**Video**
In this video I created function, but diagnostics updated only after a long time. In all IDE's diagnostics updated immediately even without save.
[![asciicast](https://asciinema.org/a/sg3YEh2gWjnXkHE0PD5hUJqZV.svg)](https://asciinema.org/a/sg3YEh2gWjnXkHE0PD5hUJqZV)

**Comment by hrsh7th** [comment]
Sorry, the experimental flags are checked existence only.

Thank you for create repro steps.
I try to run cmake but it failed on my mac.
I will try to check on any Linux environment.

**Comment by hrsh7th** [comment]
To prepare the Linux environment take time So I try to detect the cause by removing `updating condition for diagnostics` one by one...

I was remove one condition for now on the master branch.

**Comment by hrsh7th** [comment]
My suspicion points are

- Bug of diff computing algorithm
    - We should check with replacing vim-lsc's diff algorithm instead.

- Bug of condition for updating/skipping diagnostics
    - We should check with diagnostics passthrough mode

**Comment by Shatur** [comment]
> I try to run cmake but it failed on my mac.
> I will try to check on any Linux environment.

Sorry, I did now know that you use Mac. Provided program is Linux-only. Try the following minimal example (try to create `void test2()` function in `keyboarddaemon.h` and it implementation in `keyboarddaemon.h`):
[Sample.zip](https://github.com/hrsh7th/vim-lamp/files/4655111/Sample.zip)

> I try to detect the cause by removing updating condition for diagnostics one by one...

Just tested latest master, have the same issue. But why remove? I thought that new condition should be **added**. The diagnostics just not updated properly after creating function definition. I think in is because of different files. If I will create definition in the same file (C++ have such possibility) - everything works.

> My suspicion points are

I agree. I think that there is two problems. Currently I did not encountered the diff bug. But the repro example shows that diagnostics do not updated with different files.

Could you explain what is passthrough mode?

**Comment by hrsh7th** [comment]
Thank you for Sample.zip! I try to reproduce it!

The diagnostics feature is too difficult.

1. LSP client should send `text changes` to the server per press key.
2. LSP server `typically` computes diagnostics every `text changes` and send it to client.
3. LSP client shouldn't show every received diagnostic to improve UX.

For example, vim-lamp skip or debounce diagnostics when these cases.

1. [SKIP] The target buffer does not visible on any window.
2. [DEBOUNCE] In insert-mode.
3. and some conditions...

So I suspect the cause is in these conditions.

**Comment by hrsh7th** [comment]
> Could you explain what is passthrough mode?

Passthrough mode is removing the some above conditions on diagnostics feature.

**Comment by Shatur** [comment]
Thanks for explanations!
Yes, I tested with passthrough mode and the issue still exists. 
`vim-lsp` also have the same bug.

**Comment by hrsh7th** [comment]
Thank you for your investigate.

Oh...
If vim-lsp has the same problem, this problem caused by the server maybe.

Anyway, I'll try to detect the cause with your example project!

**Comment by Shatur** [comment]
Thanks for looking into it!

**Comment by hrsh7th** [comment]
Unfortunately, It seems the clangd's problem.
I searching to solve this on the clangd's issue tracker.

**Comment by Shatur** [comment]
Bad to hear :( Other IDE's that use `clangd` do not have this issue.

**Comment by hrsh7th** [comment]
Hm...
I try to check with coc.nvim temporary but it has same problem...

**Comment by hrsh7th** [comment]
I try again with VSCode.

**Comment by hrsh7th** [comment]
VSCode has same issue...

**Comment by Shatur** [comment]
Let me check it myself :)

**Comment by hrsh7th** [comment]
Yes, thanks!

FYI: I used `vscode-clangd`.


Currently, My thoughts are

- Your problem cause by the clangd.
- My problem is a more rare case of bug of the diagnostics maybe.
- I want to improve the behavior of diagnostics on clangd if we can.

**Comment by Shatur** [comment]
Wow, you are right. Only QtCreator works with this correctly (without save).
But VSCode updates this file immediately after "Save all" and typing any symbol. Maybe is it possible to at least achieve something similar?

**Comment by Shatur** [comment]
Yes, I think that my and your problems is different. If I reproduce your problem with minimal example - I will let you know.

**Comment by hrsh7th** [comment]
Anyway, I re-enable diagnostics and incremental diff.
when you met the problem, please tell me it.

And I try to check the ccls.

**Comment by Shatur** [comment]
Did I understand correctly that incremental diff is now enabled by default and "Passthrough mode" is disabled by default?

**Comment by hrsh7th** [comment]
Yes. You are right.

**Comment by Shatur** [comment]
But I still see [checks](https://github.com/hrsh7th/vim-lamp/blob/6994cf6c317e8f9f77ded3e21469f5db097a9103/autoload/lamp/view/diff.vim#L7) in the latest master. Is it expected?

**Comment by hrsh7th** [comment]
Sorry. I've updated it `now`. 🙇

**Comment by Shatur** [comment]
Thanks!
Is it possible to to fix diagnostics at least after saving all files as in VSCode?

**Comment by hrsh7th** [comment]
I try it.

I guess VSCode sent `didChange` on save (if the server does not support `textDocumentSync.save` capability.

**Comment by Shatur** [comment]
It would be awesome. In C++ is very common to use a lot of different files :)

**Comment by hrsh7th** [comment]
Could you test https://github.com/hrsh7th/vim-lamp/commit/81a9312c00b96b02f58d78fea40e6544115648c1 ?

**Comment by Shatur** [comment]
Any differences with and without this fix :(
Diagnostics becomes updated only after several time.

**Comment by Shatur** [comment]
[![asciicast](https://asciinema.org/a/uBRk29M7hmPm3i16faFfqDlkm.svg)](https://asciinema.org/a/uBRk29M7hmPm3i16faFfqDlkm)

**Comment by Shatur** [comment]
In this video I used small helper command that just creates implementation in `.cpp`. But the issue occurs even with manual editing.

**Comment by hrsh7th** [comment]
Sorry. Probably I'm misunderstood your mean.

I just implemented updating diagnostics at saving the buffer.

![Kapture 2020-05-21 at 0 13 27](https://user-images.githubusercontent.com/629908/82463698-17cb7c80-9af8-11ea-8d9c-8d2e8b368f8f.gif)

**Comment by Shatur** [comment]
Tested, but it looks like I still have the same issue as with creating implementation:
[![asciicast](https://asciinema.org/a/9ahyVvMC7GgvmHopi0qjQPJPN.svg)](https://asciinema.org/a/9ahyVvMC7GgvmHopi0qjQPJPN)

**Comment by hrsh7th** [comment]
Is the following feature you want?

1. Open `*.cpp` file and `*.h` file.
2. Edit `*.h` file (And the edit occurs error on `*.cpp`.
3. Focus `*.cpp`.
4. Update `*.cpp` diagnostics.

If so, I can implement it (But it is not general implementation, I think).

**Comment by Shatur** [comment]
Yes, it would be just awesome. But if this is a bit hacky I would suggest to just add the ability to update manually reload diagnostics. This would also help to debug your issue.

**Comment by hrsh7th** [comment]
I've added `experimental.did_change_on_focus` option.
Please test it.

**Comment by Shatur95** [comment]
How can I enable it?

**Comment by Shatur** [comment]
Sorry, but how can I enable it?

**Comment by Shatur95** [comment]
Hm... I experienced another issue:
After swathing between header / cpp completion sometimes stop to working. I tested previous commits (via checkout) and the issue still exits.

**Comment by Shatur** [comment]
Hm... I experienced another issue:
After swathing between header / cpp completion sometimes stop to working. I tested previous commits (via checkout) and the issue still exits:
[![asciicast](https://asciinema.org/a/Xhmk2C0ttJXvqgGxjCsPd3KSJ.svg)](https://asciinema.org/a/Xhmk2C0ttJXvqgGxjCsPd3KSJ)

**Comment by hrsh7th** [comment]
Sorry. You can enable it by `lamp#config('experimental.did_change_on_focus', v:true)` on lamp#initialized autocmd.

I try to use this feature more.

**Comment by Shatur** [comment]
> Sorry. You can enable it by lamp#config('experimental.did_change_on_focus', v:true) on lamp#initialized autocmd.

Thanks, tested, but still not works :(
[![asciicast](https://asciinema.org/a/bNoJroAMR8BsF4OAIR7EB5e6P.svg)](https://asciinema.org/a/bNoJroAMR8BsF4OAIR7EB5e6P)

**Comment by Shatur** [comment]
> Hm... I experienced another issue:
After swathing between header / cpp completion sometimes stop to working. I tested previous commits (via checkout) and the issue still exits.

This is another issue that related with recent changes. I found a way how to reproduce it:
1. Open .cpp file. You will have completion.
2. Go to the end of file and type `void`.
3. Delete it.
4. Completion from LSP stop working.

Video:
[![asciicast](https://asciinema.org/a/zWLmyhTZbvthAaRbz35kGSzVW.svg)](https://asciinema.org/a/zWLmyhTZbvthAaRbz35kGSzVW)

**Comment by hrsh7th** [comment]
> Thanks, tested, but still not works :(

Very strange...
My environment works well...



> This is another issue that related with recent changes. I found a way how to reproduce it:
> 
> Open .cpp file. You will have completion.
> Go to the end of file and type void.
> Delete it.
> Completion from LSP stop working.
> Video:

It is reproduced on my machine.
It seems bug of lamp's diff algorithm.
It is difficult to fix but I try it.

**Comment by Shatur** [comment]
> Very strange...
> My environment works well...

Okay, I will try without any plugins.

> It is reproduced on my machine.
> It seems bug of lamp's diff algorithm.
> It is difficult to fix but I try it.

At least it reproduced :D

**Comment by hrsh7th** [comment]
```
16:43:16	clangd      	-> [NOTIFY]	{'method': 'textDocument/didChange', 'params': {'contentChanges': [{'range': {'end': {'character': 0, 'line': 12}, 'start': {'character': 0, 'line': 12}}, 'text': '', 'rangeLength': 0}], 'textDocument': {'uri': 'file:///path/to/Sample/keyboarddaemon.cpp', 'version': 5}}}
16:43:16	clangd      	[STDERR]	I[16:43:16.873] <-- textDocument/didChangeE[16:43:16.874] Failed to update /path/to/Sample/keyboarddaemon.cpp: Line value is out of range (12)
```

Probably end of line diff was broken.

**Comment by hrsh7th** [comment]
I found the critical bug...

**Comment by Shatur** [comment]
> Okay, I will try without any plugins.

Tested `  call lamp#config('experimental.did_change_on_focus', v:true)
`, still not works :(

**Comment by Shatur** [comment]
I think that LSP implementation in Vimscript is hard and such thing should be native. I tested a lot of lsp plugins and your seems the best to me because of speed. But now native implementation becomes usable with [completion-nvim](https://github.com/haorenW1025/completion-nvim) and [diagnostics-nvim](https://github.com/haorenW1025/diagnostic-nvim), I will switch to it for now for daily use. I suggest to check it out :) But feel free to request to test something with vim-lamp.

**Comment by hrsh7th** [comment]
Please forgive my lack of strength.

Thank you for interested to vim-lamp.

My plugin's supports vim and nvim both.
But I wondering to drop vim support is good for now...the reason is that we should use lua maybe...

**Comment by Shatur** [comment]
No need to apologize, you did an awesome job!
Switching to lua sounds as a good idea as for me. But Neovim will have native LSP implementation. It very flexible and fast. Maybe it would be better to use it?

**Comment by hrsh7th** [comment]
I'm using multiple language server for one filetype every day but native lsp does not support it in my knowledge.

I planning to create completion plugin in lua.

**Comment by Shatur** [comment]
> I'm using multiple language server for one filetype every day but native lsp does not support it in my knowledge.

Interesting :) Why do you need several language servers?

> I planning to create completion plugin in lua.

You will discontinue `compete` or just rewrite it? Curretnly `compete` is very good, I tested `completion-nvim` but it lacks some features. I thought to request native LSP completion source for compete (if this is possible).
But implementation in lua can be much faster, you are right.

**Comment by hrsh7th** [comment]
For example, Editing `typescript`, Probably we choose `typescript-language-server` but I want to use `prettier` for formatting.

In this case, we use `typescript-language-server` and `diagnostic-languageserver` or `efm-langserver` both.

Another case, Editing `Vim script`, we choose `vim-language-server` but I want to use `vint` for linting.

I love `compete`'s behavior and implementation but I do not satisfied with its performance.
So I planning to improve with lua.

I'm considering whether to modify or scratch.

**Comment by Shatur** [comment]
Did not know about such case. Interesting. I using `vint` manually for now :(

It would be awesome :)

**Comment by hrsh7th** [comment]
When using vint with diagnostic-languageserver, we will get vint diagnostics on realtime :)

**Comment by Shatur** [comment]
Sounds awesome. I hope such ability will be added to the native LSP plugin.

I followed you on github, to track updates about `compete` rewriting.

**Comment by hrsh7th** [comment]
We can reload diagnostics by BufWritePost.

**Comment by Shatur** [comment]
Thanks, did not know :)
But diff problem still exists?

**Comment by hrsh7th** [comment]
Probably the diff problem was solved (I have some internal improvement point for it but it does not problem for now).

**Comment by Shatur** [comment]
I tested the latest commit and the diff problem looks solved!

But `:doautocmd BufWritePost` not reload diagnostics :(
[![asciicast](https://asciinema.org/a/z0WQdcUYju1hrLqbjDLXFFr6k.svg)](https://asciinema.org/a/z0WQdcUYju1hrLqbjDLXFFr6k)

I have another idea to solve it. What if just improve diagnostics?
I used [diagnostics-nvim](https://github.com/haorenW1025/diagnostic-nvim) which adds some nice collbacks for native LSP. The idea is hide annoying inline diagnostics and just display a sign. And user will can show it manually. This also better for multiline errors (very common in C++).
Is it possible to implement such diagnostics in this plugin? This is how VSCode works and it will remove the need for a manual reset of the diagnostic since does not interfere.

**Comment by Shatur** [comment]
Example:
![изображение](https://user-images.githubusercontent.com/22453358/82738436-2c01ba80-9d40-11ea-9aa6-f0d218b420dc.png)

**Comment by Shatur** [comment]
@hrsh7th, I see that make diagnostics look in this way. Thank you a lot! Looks very nice.

**Comment by hrsh7th** [comment]
Thank you for request it.

I was thinking to implement it but dosen't because bit troublesome.

The diagnostics visualization is hard thing so I want to your feedback.
